fdbserver/VersionedBTree.actor.cpp

```diff
@@ -8334,6 +9021,164 @@ TEST_CASE(":/redwood/correctness/pager/cow") {
+
+		LogicalPageID extID = pager->newLastExtentID();
+		m_extentQueue.create(pager, extID, "ExtentQueue", pager->newLastQueueID(), true);
+		pager->pushExtentUsedList(m_extentQueue.queueID, extID);
+
+		state int v;
+		state ExtentQueueEntry<16> e;
+		generateRandomData(e.entry, 16);
+		state int sinceYield = 0;
+		for (v = 1; v <= numEntries; ++v) {
+			// Sometimes do a commit
+			if (currentCommitSize >= targetCommitSize) {
+				printf("currentCommitSize: %d, cumulativeCommitSize: %d, pageCacheCount: %d\n",
+				       currentCommitSize,
+				       cumulativeCommitSize,
+				       pager->getPageCacheCount());
+				wait(m_extentQueue.flush());
+				wait(pager->commit());
+				cumulativeCommitSize += currentCommitSize;
+				targetCommitSize = deterministicRandom()->randomInt(2e6, 30e6);
+				currentCommitSize = 0;
+			}
+
+			// push a random entry into the queue
+			m_extentQueue.pushBack(e);
+			currentCommitSize += 16;
+
+			// yield periodically to avoid overflowing the stack
+			if (++sinceYield >= 100) {
+				sinceYield = 0;
+				wait(yield());
+			}
+		}
+		cumulativeCommitSize += currentCommitSize;
+		printf(
+		    "Final cumulativeCommitSize: %d, pageCacheCount: %d\n", cumulativeCommitSize, pager->getPageCacheCount());
+		wait(m_extentQueue.flush());
+		extentQueueState = m_extentQueue.getState();
+		printf("Commit ExtentQueue getState(): %s\n", extentQueueState.toString().c_str());
+		pager->setMetaKey(extentQueueState.asKeyRef());
+		wait(pager->commit());
+
+		Future<Void> onClosed = pager->onClosed();
+		pager->close();
+		wait(onClosed);
+	}
+
+	printf("Reopening pager file from disk.\n");
+	pager = new DWALPager(pageSize, pagesPerExtent, fileName, cacheSizeBytes, remapCleanupWindow, concExtentReads);
+	wait(success(pager->init()));
+
+	printf("Starting ExtentQueue FastPath Recovery from Disk.\n");
+
+	// reopen the pager from disk
+	state Key meta = pager->getMetaKey();
+	memcpy(&extentQueueState, meta.begin(), meta.size());
+	extentQueueState.fromKeyRef(meta);
+	printf("Recovered ExtentQueue getState(): %s\n", extentQueueState.toString().c_str());
+	m_extentQueue.recover(pager, extentQueueState, "ExtentQueueRecovered");
+
+	state double intervalStart = timer();
+	state double start = intervalStart;
+	state Standalone<VectorRef<LogicalPageID>> extentIDs = wait(pager->getUsedExtents(m_extentQueue.queueID));
+
+	// fire read requests for all used extents
+	state int i;
+	for (i = 1; i < extentIDs.size() - 1; i++) {
+		LogicalPageID extID = extentIDs[i];
+		pager->readExtent(extID);
+	}
+
+	state PromiseStream<Standalone<VectorRef<ExtentQueueEntry<16>>>> resultStream;
+	state Future<Void> queueRecoverActor;
+	queueRecoverActor = m_extentQueue.peekAllExt(resultStream);
+	state int entriesRead = 0;
+	loop choose {
+		when(Standalone<VectorRef<ExtentQueueEntry<16>>> entries = waitNext(resultStream.getFuture())) {
+			entriesRead += entries.size();
+			if (entriesRead == m_extentQueue.numEntries)
+				break;
+		}
+	}
+
```

Comment:
Note:  This comment applies to this loop and the other read loop in recover.

If a bug or corruption causes `numEntries` to be wrong, this loop will either stop early and not realize it didn't read everything or it will wait too long if it expects too many entries.

I think it would be better for `peekAll()` to send `end_of_stream()` at completion, and before doing so validate that the amount of items sent matches the queue metadata.  If not, it should throw an error.

The read loop would then be
```
try {
  loop {
    choose {
      when(// next item) {
         // process results 
      }
      when(wait(queueRecoverActor)) {
         queueRecoverActor = Never();
      }
  }
} catch(Error e) {
  if(e.code() != error_code_end_of_stream) {
    throw;
  }
}
```

The purpose of waiting on the actor is to catch any errors it throws, such as io errors when waiting on extent reads, which would not otherwise be surfaced as they do not go to the result stream.

The purpose of setting the peek future to Never after it is ready instead of just breaking is to avoid a race where the actor return is delivered prior to the final result send (which won't occur as written but it could if a wait were introduced in the process results code block).

With this form, any queue data error detected such as the entry count being wrong could technically be either thrown or sent to the output stream, but I think throwing is more natural.

Reply:
Thanks for catching this and explaining in great detail. And saving a future headache :)
I have made the suggested changes.